src/sp/it/pl/audio/tagging/MetadataReader.java

```diff
@@ -37,7 +37,7 @@ public static Metadata readMetadata(Item item) {
 		}
 
 		if (item.isFileBased()) {
-			return readAudioFile(item.getFile())
+			return readAudioFile(noNull(item.getFile()))
```

Comment:
You already check whether it isFileBased, so `getFile` doubles it up again. Either create a ``getFileDirect` method, or create some lambda function that checks whether the file isFileBased and executes one of two lambdas depending on the result with the file as parameter.

Reply:
Java... Not even going to bother. Will be refactored.

But I have to say:
The proposed lambda is an elegant solution and much more viable than the proposed method, but its still a complication that is not substantiated. I hate to call 'premature optimization', but calling same method twice is not as bad as it sounds - as long as its instantaneous call and we stay in Java-land. I also hate calling the same thing twice, but please bear in mind that isFileBased being called in getFile() is implementation detail that the caller should not be concerned about. Just let the API do its job instead of trying to outsmart it.

Readability beats performance here.

Reply:
PS: the (noninlined) lambda would probably incur more performance impact than calling isFileBased() twice... Using callbacks to hide internal logic is fine, using it for convenience is questionable practice. API with callbacks are suspicious.

Reply:
Of course you'd inline that ^^